stan/math/rev/functor/ode_bdf_adjoint.hpp

```diff
@@ -0,0 +1,182 @@
+ *   template<typename T_t, typename T_y, typename... T_Args>
+ *   Eigen::Matrix<stan::return_type_t<T_t, T_y, T_Args...>, Eigen::Dynamic, 1>
+ *     operator()(const T_t& t, const Eigen::Matrix<T_y, Eigen::Dynamic, 1>& y,
+ *     std::ostream* msgs, const T_Args&... args);
+ *
+ * t is the time, y is the state, msgs is a stream for error messages, and args
+ * are optional arguments passed to the ODE solve function (which are passed
+ * through to \p f without modification).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_0 Type of initial time
+ * @tparam T_ts Type of output times
+ * @tparam T_Args Types of pass-through parameters
+ *
+ * @param function_name Calling function name (for printing debugging messages)
+ * @param f Right hand side of the ODE
+ * @param y0 Initial state
+ * @param t0 Initial time
+ * @param ts Times at which to solve the ODE at. All values must be sorted and
+ *   not less than t0.
+ * @param relative_tolerance Relative tolerance passed to CVODES
+ * @param absolute_tolerance Absolute tolerance passed to CVODES
+ * @param max_num_steps Upper limit on the number of integration steps to
+ *   take between each output (error if exceeded)
+ * @param[in, out] msgs the print stream for warning messages
+ * @param args Extra arguments passed unmodified through to ODE right hand side
+ * @return Solution to ODE at times \p ts
+ */
+template <typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args, require_eigen_col_vector_t<T_y0>* = nullptr>
+std::vector<Eigen::Matrix<stan::return_type_t<T_y0, T_t0, T_ts, T_Args...>,
+                          Eigen::Dynamic, 1>>
+ode_bdf_adjoint_tol_impl(const char* function_name, const F& f, const T_y0& y0,
+                         const T_t0& t0, const std::vector<T_ts>& ts,
+                         double relative_tolerance, double absolute_tolerance,
+                         long int max_num_steps, std::ostream* msgs,
+                         double absolute_tolerance_B,
+                         double absolute_tolerance_QB,
+                         long int steps_checkpoint, const T_Args&... args) {
+  /*
+  const auto& args_ref_tuple = std::make_tuple(to_ref(args)...);
+  return apply(
+      [&](const auto&... args_refs) {
+        auto integrator
+            = new stan::math::cvodes_integrator_adjoint_vari<CV_BDF, F, T_y0,
+  T_t0, T_ts, T_Args...>( function_name, f, y0, t0, ts, relative_tolerance,
+  absolute_tolerance, max_num_steps, msgs, args_refs...); return
+  (*integrator)();
+      }, args_ref_tuple);
+  */
+  auto integrator
+      = new stan::math::cvodes_integrator_adjoint_vari<CV_BDF, F, T_y0, T_t0,
+                                                       T_ts, T_Args...>(
+          function_name, f, y0, t0, ts, relative_tolerance, absolute_tolerance,
+          absolute_tolerance_B, absolute_tolerance_QB, steps_checkpoint,
+          max_num_steps, msgs, args...);
```

Comment:
Can you change this to:

```
new stan::math::cvodes_integrator_adjoint_vari<CV_BDF, F, T_y0, T_t0,
                                                       T_ts, T_Args...>(
          function_name, f, eval(y0), t0, ts, relative_tolerance, absolute_tolerance,
          absolute_tolerance_B, absolute_tolerance_QB, steps_checkpoint,
          max_num_steps, msgs, eval(args)...)
```

Reply:
Oh, same problem here, cause `T_Args` ... Hmm. Can always do ugly things like `decltype(eval(std::declval<const T_Arg&>()))...` to get the types after the const ref T_Args have been passed through `eval`.

Reply:
But probably better on the inside to make it so that the constructor here can digest expressions properly. Sounds like this wasn't an easy change so I'll have a look in  asec.

Reply:
Thanks...but this results in the same compile error.

Reply:
Which test you building? Checkout out the code now

Reply:
Or do I need to be building the Stan model to see this error?

Reply:
I'm not able to build the Stan model.

Reply:
This is my next guess as to the code to use (eval returns plain types and I forgot that the type of y0 might change last time):

```c++
auto integrator
    = new stan::math::cvodes_integrator_adjoint_vari<CV_BDF, F, plain_type_t<T_y0>, T_t0,
                                                     T_ts, plain_type_t<T_Args>...>
    (function_name, f, eval(y0), t0, ts, relative_tolerance, absolute_tolerance,
     absolute_tolerance_B, absolute_tolerance_QB, steps_checkpoint,
     max_num_steps, msgs, eval(args)...);
```

Can you post the full error message with this code if it's not easy to post the code?

Reply:
wait, I take it back. I'm able to build the Stan model.

Reply:
it compiles now and the model diagnostic look good. Do we need to review `reduce_sum` code for this?

Reply:
`reduce_sum` is tested with the expressions tests (https://github.com/stan-dev/math/blob/develop/test/expressions/expression_test_helpers.hpp#L209) so we should be fine there. 

All exposed functions that have Eigen typed args are run through the expression tests. Before this goes in, the expression tests will need to be handle this as well. Expressions tests are generated automatically but need a bit manual help for variadic stuff. But should be easy once we get there.